rewrite-groovy/src/main/java/org/openrewrite/groovy/GroovyParserVisitor.java

```diff
@@ -2502,27 +2512,52 @@ private String sourceSubstring(int beginIndex, String untilDelim) {
      * @return the level of parenthesis parsed from the source
      */
     private int determineParenthesisLevel(int childLineNumber, int parentLineNumber, int childColumn, int parentColumn) {
-        int saveCursor = cursor;
-        whitespace();
-        int untilCursor = cursor;
-        if (childLineNumber > parentLineNumber) {
-            for (int i = 0; i < (childLineNumber - parentLineNumber); i++) {
-                untilCursor = source.indexOf('\n', untilCursor) + 1; // +1; set cursor past `\n`
-            }
-            untilCursor += childColumn - 1; // -1; skip previous `\n`
+        // Map the coordinates
+        int startingLineNumber;
+        int startingColumn;
+        int endingLineNumber;
+        int endingColumn;
+        if (childLineNumber == parentLineNumber) {
+            startingLineNumber = childLineNumber;
+            endingLineNumber = childLineNumber;
+            startingColumn = Math.min(childColumn, parentColumn);
+            endingColumn = Math.max(childColumn, parentColumn);
+        } else if (childLineNumber > parentLineNumber) {
+            startingLineNumber = parentLineNumber;
+            endingLineNumber = childLineNumber;
+            startingColumn = parentColumn;
+            endingColumn = childColumn;
         } else {
-            untilCursor += childColumn - parentColumn;
+            startingLineNumber = childLineNumber;
+            endingLineNumber = parentLineNumber;
+            startingColumn = childColumn;
+            endingColumn = parentColumn;
+        }
+
+        // Grab from the source the current part
+        StringBuilder result = new StringBuilder();
+        for (int i = startingLineNumber; i <= endingLineNumber; i++) {
+            if (i == startingLineNumber && i == endingLineNumber) {
+                result.append(source, sourceByLineNumberOffsets[i - 1] + startingColumn - 1, sourceByLineNumberOffsets[i - 1] + endingColumn - 1);
+            } else if (i == startingLineNumber) {
+                result.append(source, sourceByLineNumberOffsets[i - 1] + startingColumn - 1, sourceByLineNumberOffsets[i]);
+            } else if (i == endingLineNumber) {
+                result.append(source, sourceByLineNumberOffsets[i - 1], sourceByLineNumberOffsets[i - 1] + endingColumn - 1);
+            } else {
+                int end = sourceByLineNumberOffsets.length == i ? source.length() : sourceByLineNumberOffsets[i];
+                result.append(source, sourceByLineNumberOffsets[i - 1], end);
+            }
         }
+
+        // Determine level of parentheses
         int count = 0;
-        for (int i = cursor; i < untilCursor; i++) {
-            if (source.charAt(i) == '(') {
+        for (char ch : result.toString().toCharArray()) {
```

Comment:
`toCharArray()` should almost always be avoided and instead `charAt()` should be used.